### PR TITLE
Implement uniform distribution of events in N sphere in FakeMD

### LINKED ISSUE #33213: Bug in FakeMDEventData for spherically symmetric Gaussian peak
**Describe the bug**
`FakeMDEventData` is used to generate fake peaks of backgrounds for e.g. tests of integration algorithms, sliceviewer plotting etc.
Here I compare what should be the same spherically symmetric Gaussian peak produced in two ways: LHS uses the built-in `PeakParams` argument, the RHS plot used a more general input called `EllipsoidParams` to specify a peak with a diagonal covariance matrix (eigenvectors and eigenvalues passed explicitly).

image

There is a slight difference expected because the `PeakParams` spherical gaussian is passed a radius (I've assumed this is 3 sigma but it's not in the algorithm docs) - and it does not add any events outside this radius. But it can clearly be seen that the peak is not spherically symmetric.

**To Reproduce**
(1) Create data and peaks
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)
FakeMDEventData(ws, EllipsoidParams='1e+05,0,1,0,1,0,0,0,1,0,0,0,1,0.01,0.01,0.01,1', RandomSeed='3873875')
p = CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
SetUB(p, 1,1,2,90,90,120)
pk = p.createPeakHKL([0,1,0]) # axes aligned with viewing axes
p.addPeak(pk)

FakeMDEventData(ws, PeakParams='1e+05,0,1,1,0.3', RandomSeed='3873875')
pk = p.createPeakHKL([0,1,1]) # axes aligned with viewing axes
p.addPeak(pk)
```
(2) Open ws in sliceviewer
(3) Overlay peaks workspace
(4) Double click on first and second peak to navigate to the correct region

**Expected behavior**
Both peaks should be spherically symmetric, but the second one (created using `PeakParams`) does not.

**Platform/Version (please complete the following information):**
 - OS: presumably all but found on windows
 - Mantid Version 6.2 nightly from 4 Jan

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Previously the orientation/direction of a unit vector was found from uniformly distributed random coords in a hypercube, which were then used to generate an event inside a hypershpere. However this gives greater probability of events near the vertices leading to lines of intensity pointing towards the vertices. Now the orientation is determined from uniform sampling of surface of unit N-sphere (algorithm adapted from [1]).

image

[1] http://corysimon.github.io/articles/uniformdistn-on-sphere/

**To test:**

(1) Run this script
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,1,90,90,90)
p = CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
FakeMDEventData(ws, PeakParams='1e+05,0,1,1,0.3', RandomSeed='3873875')
pk = p.createPeakHKL([0,1,1]) # axes aligned with viewing axes
p.addPeak(pk)
```
(2) View in `ws` in sliceviewer
(3) Overlay `peaks`
(4) Double click on first row in peak viewer table on RHS
(5) Should be region of random uniform events within sphere of radius 0.3

Fixes #33213 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DannyHindson** [comment]
@RichardWaiteSTFC - I had a look through the changes you did here which are the main part of this PR and they look fine. Would you be happy to look over the final few commits I did to get this over the line? Just swapping the Kernel implementation of normal_distribution in for the std version and some test changes. If it all looks OK I'll approve it. Don't think it's necessary for someone else to review

**Comment by RichardWaiteSTFC** [comment]
> @RichardWaiteSTFC - I had a look through the changes you did here which are the main part of this PR and they look fine. Would you be happy to look over the final few commits I did to get this over the line? Just swapping the Kernel implementation of normal_distribution in for the std version and some test changes. If it all looks OK I'll approve it. Don't think it's necessary for someone else to review

Thanks @DannyHindson for getting it over the line. I'll have a look and give it a quick test this morning.

File: Framework/DataObjects/src/FakeMD.cpp [hunk 1]
```diff
@@ -204,7 +205,7 @@ template <typename MDE, size_t nd> void FakeMD::addFakeEllipsoid(typename MDEven
 
   // prepare random number generator
   std::mt19937 rng(static_cast<unsigned int>(m_randomSeed));
-  std::normal_distribution<double> d(0.0, 1.0); // mean = 0, std = 1
+  Kernel::normal_distribution<double> d(0.0, 1.0); // mean = 0, std = 1
```
Comment (RichardWaiteSTFC) [inline Framework/DataObjects/src/FakeMD.cpp:208]: Good spot that `normal_distribution` is also called in the ellipsoid peak simulation - this is probably the root of inconsistent behaviour between OS in ellipsoid tests for `IntegratePeaksMD` that I've struggled with in the past.

**Review by RichardWaiteSTFC — COMMENTED** [review]
Looks good and functions as expected. Don't think it needs release notes as it's a dev only algorithm used for testing.

**Review by DannyHindson — APPROVED** [review]

**Pull request merged**